### PR TITLE
refactor(store): simplePatch

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?

This commit updates, where the for-in loop was used, resulting in copying from prototype to own properties, thus consuming more memory. Prior to NodeJS 12 (V8 versions before 7.4) there was an optimization that was improving the situation and since that logic was removed in favor of other optimizations, the situation with memory consumption caused by the for-in loop got worse. This commit adds a check to make sure we copy only own properties over to cloned object.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by arturovt** [comment]
Could you, please, provide benchmarks that it’s really faster on most browsers? I’m onto phone and I cannot check it.
Also you’re talking about nodejs12, but our code is running in browser. Could you also, please, create a spec that shows that for-in consumes a lot of memory? I started googling and didn’t find anything unfortunately:(

**Comment by splincode** [comment]
http://jsben.ch/x4Ig4

image

I understand that these are not huge results, however, we can still help improve performance

**Comment by arturovt** [comment]
image

**Comment by splincode** [comment]
When you're using for (var key in obj) it will loop through the given object + its parent objects' properties on the prototype chain until it reaches the end of the chain. As you want to check only specific object's properties, you need to use hasOwnProperty.

**Comment by arturovt** [comment]
I think this is micro optimization. As you see I just gave you the screen from my phone.
We have to focus mobile devices first, not PCs.

**Comment by splincode** [comment]
@markwhitfeld updated

**Comment by splincode** [comment]
I recently saw in our company how people do this and it terrified me, so I decided to make this pull request

```ts
ctx.patchState(new MyClass('hello world'));
```

File: packages/store/src/internal/state-operators.ts [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
-    if (isPrimitive) {
-      throw new Error('Patching primitives is not supported.');
-    }
-    const newState = { ...(<any>existingState) };
-    for (const k in val) {
-      newState[k] = val[k];
-    }
-    return <T>newState;
+
+    return { ...existingState, ...val };
```
Comment (markwhitfeld) [inline packages/store/src/internal/state-operators.ts]: Technically this change has the potential to be a breaking change for someone....?

File: packages/store/tests/state.spec.ts [hunk 2]
```diff
@@ -348,20 +348,45 @@ describe('State', () => {
...
+      expect(resultState.y).toEqual(undefined);
+    });
+
     it('should throw exception if value is array', () => {
       try {
         const simple: string[] = ['hello'];
         simplePatch(['world'])(simple);
       } catch (e) {
-        expect(e.message).toEqual('Patching arrays is not supported.');
+        expect(e.message).toEqual(MESSAGES[CODE.PATCHING_ARRAY]());
```
Comment (markwhitfeld) [inline packages/store/tests/state.spec.ts]: Tests should still be explicit about the message. It should not look up the expected value.
```suggestion
        expect(e.message).toEqual('Patching arrays is not supported.');
```

Comment (splincode) [reply packages/store/tests/state.spec.ts]: image

I see that we adhere to the same errors everywhere for verification

Comment (markwhitfeld) [reply packages/store/tests/state.spec.ts]: This is an anti-pattern in tests that we should avoid.
Your expectation should always be constant and not calculated.

File: packages/store/src/internal/state-operators.ts (hunk 1 above)
Comment (splincode) [reply packages/store/src/internal/state-operators.ts]: not sure, so rolled back

File: packages/store/src/internal/state-operators.ts [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
-      throw new Error('Patching primitives is not supported.');
-    }
-    const newState = { ...(<any>existingState) };
-    for (const k in val) {
-      newState[k] = val[k];
+
+    const newState: any = { ...(existingState as any) };
+    for (const key in val) {
+      // deep clone for patch compatibility
+      // noinspection JSUnfilteredForInLoop (IDE)
```
Comment (arturovt) [inline packages/store/src/internal/state-operators.ts]: Could you remove this please?

Comment (splincode) [reply packages/store/src/internal/state-operators.ts]: my IDEA (WebStorm) swears red constantly

**Pull request merged**